web/src/apps/selector.ts

```diff
@@ -0,0 +1,179 @@
+import { apply, css, tw } from '@twind/core'
+import { LitElement, html } from 'lit'
+import { customElement, property, state } from 'lit/decorators.js'
+import { Ref, createRef, ref } from 'lit/directives/ref.js'
+
+import { withTwind } from '../twind'
+import type { DocumentView } from '../types'
+/**
+ * This class extends LitElement to include a tw instance.
+ */
+class TWLitElement extends LitElement {
+  protected tw: typeof tw
+}
+
+/**
+ * UI selector
+ *
+ * A selector that updates some display portion of the UI
+ */
+@customElement('stencila-ui-selector')
+@withTwind()
+export class UISelector extends TWLitElement {
+  /**
+   * Ref to allow us to close the details element when needed.
+   */
+  detailsRef: Ref<HTMLDetailsElement> = createRef()
+  /**
+   * Manages the open state of the open listbox
+   */
+  @state()
+  private open: boolean = false
+  /**
+   * Label displayed when listbox is not open
+   */
+  @property()
+  label: string = ''
+  /**
+   * List of values to render in the list
+   */
+  @property({ type: Array })
+  list: [string, string][] = []
+  /**
+   * Event to call when a list element is selected
+   */
+  @property()
+  clickEvent: (e: Event) => void | undefined
+  /**
+   * Target property in parent component to evaluate
+   */
+  @property()
+  target: DocumentView | string
+  override render() {
+    return html`
+      ${this.renderOverlay()}
+      <details
+        role="list"
+        class="group text-gray-aluminium open:text-brand-blue open:border-b-brand-blue p-0 relative block flex-grow open:z-50"
+        ${ref(this.detailsRef)}
+      >
+        ${this.renderSummary()} ${this.renderList()}
+      </details>
+    `
+  }
+
+  private renderSummary() {
+    const styles = apply([
+      'text-base font-bold',
+      'leading-none',
+      'select-none',
+      'appearance-none ',
+      'min-w-fit',
+      'py-2 px-4',
+      'bg-white',
+      'border-b-4 border-b-transparent',
+      'transition-all ease-in-out',
+      'flex',
+      'items-center',
+      'group-hover:text-brand-blue group-hover:border-b-brand-blue',
+    ])
+
+    const hideMarker = css`
+      &::marker {
+        display: none;
+        font-size: 0;
+      }
+    `
```

Comment:
Love this use of `apply` and `css` to keep templates and code in general clean

Reply:
I'm going to add some docs for this I think